src/module.cpp

```diff
@@ -0,0 +1,167 @@
+    cache_type &cache, const exclusion::context_policy &exclusion,
+    const std::unordered_map<std::string, std::shared_ptr<matcher::base>> &dynamic_matchers,
+    ddwaf::timer &deadline) const
+{
+    verdict_type final_verdict = verdict_type::none;
+    for (const auto &collection : collections_) {
+        DDWAF_DEBUG("Evaluating collection: {}", collection.name);
+        auto &collection_cache = cache.collections[collection.name];
+        if (collection_cache.type >= collection.type) {
+            // If the result was cached but ephemeral, clear it. Note that this is
+            // just a workaround taking advantage of the order of evaluation of
+            // collections. Collections might be removed in the future altogether.
+            if (collection_cache.type == collection.type && collection_cache.ephemeral) {
+                collection_cache.type = core_rule::verdict_type::none;
+                collection_cache.ephemeral = false;
+            } else {
+                continue;
+            }
+        }
+
+        for (std::size_t i = collection.begin; i < collection.end; ++i) {
+            auto &rule = *rules_[i];
+            auto [event, verdict] =
+                eval_rule(rule, store, cache.rules[i], exclusion, dynamic_matchers, deadline);
+            if (event.has_value()) {
+                collection_cache.type = verdict;
+                collection_cache.ephemeral = event->ephemeral;
+
+                events.emplace_back(std::move(*event));
+                DDWAF_DEBUG("Found event on rule {}", rule.get_id());
+
+                if (verdict == verdict_type::block) {
+                    return verdict_type::block;
+                }
+
+                final_verdict = verdict_type::monitor;
+                break;
+            }
+        }
+    }
+    return final_verdict;
+}
+
+verdict_type rule_module::eval(std::vector<event> &events, object_store &store, cache_type &cache,
+    const exclusion::context_policy &exclusion,
+    const std::unordered_map<std::string, std::shared_ptr<matcher::base>> &dynamic_matchers,
+    ddwaf::timer &deadline) const
+{
+    auto &apt_deadline = get_deadline(deadline);
+
+    if (collections_.empty()) {
+        auto final_verdict = verdict_type::none;
+        for (std::size_t i = 0; i < rules_.size(); ++i) {
+            const auto &rule = *rules_[i];
+            auto &rule_cache = cache.rules[i];
+
+            auto [event, verdict] =
+                eval_rule(rule, store, rule_cache, exclusion, dynamic_matchers, apt_deadline);
+            if (event.has_value()) {
+                events.emplace_back(std::move(*event));
+                DDWAF_DEBUG("Found event on rule {}", rule.get_id());
+                final_verdict = verdict;
+                if (final_verdict == verdict_type::block) {
+                    break;
+                }
+            }
+        }
+        return final_verdict;
```

Comment:
Yes because the `null_grouping_key` results in no collections, unless I missed something (I'll double check, but the tests wouldn't pass otherwise).

A collection only allows a single match, regardless of verdict. The module allows for multiple matches in monitor mode, so they are not exactly equivalent. Admittedly I wanted to do what you're suggesting but I then realised that new business logic rules mean that multiple rules may match at once within a given module.